Add shopping cart

Dimagaa opened the pull request.

Dimagaa committed:
Add Category model
Create Category table and books-categories table

Dimagaa committed:
Add CategoryRepository and Category DTO's
Modify BookRepository and BookMapper

Dimagaa committed:
Add CategoryController and CategoryService

Dimagaa committed:
Add Pagination, Sorting, and Swagger
Implement Security Requirements

Dimagaa committed:
Add Pagination, Sorting, and Swagger
Implement Security Requirements

Dimagaa committed:
Add ShoppingCart and CartItem models

Dimagaa committed:
Add ShoppingCart and CartItem DTO's
Create Mappers

Dimagaa committed:
Implement ShoppingCartService, ShoppingCartController, ShoppingCartRepository
 Add Security
 Add swagger

Dimagaa committed:
Modify CustomGlobalExceptionHandler
Improve createCartItem() method in  ShoppingCartServiceImpl class

Dimagaa committed:
Merge branch 'main' into add_shopping_cart

Dimagaa committed:
Allow '/api/error' by anyone

Dimagaa committed:
Add logging in JwtUtil and CustomGlobalExceptionHandler classes

vovasalyha commented on src/main/java/com/app/onlinebookstore/model/CartItem.java:
```diff
@@ -0,0 +1,47 @@
+    private ShoppingCart shoppingCart;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @ManyToOne
```
make it lazy

andrii-hoienko replied:
not resolved

vovasalyha commented on src/main/java/com/app/onlinebookstore/security/JwtUtil.java:
```diff
@@ -33,19 +37,23 @@ public String generateToken(String userName) {
+            return true;
         } catch (JwtException | IllegalArgumentException e) {
-            throw new JwtException("Expired or invalid JWT token");
+            logger.error("Can't parse jwt: ", e);
```
```suggestion
            logger.warn("Expired or invalid JWT token");
```

Dimagaa committed:
Modify addCartItem(CartItemCreateRequestDto request) method in ShoppingCartServiceImpl
Delete default method bookFromId(Long id) in the BookMapper interface
Delete toModel(CartItemCreateRequestDto requestDto, ShoppingCart shoppingCart) method in the CartItemMapper class

vovasalyha commented on src/main/java/com/app/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,84 @@
+                ));
+    }
+
+    @Override
```
```suggestion
    @Override
    @Transactional
```

vovasalyha commented on src/main/java/com/app/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,84 @@
+        return cartItemMapper.toDto(cartItemRepository.save(cartItem));
+    }
+
+    @Override
```
```suggestion
    @Override
    @Transactional
```

vovasalyha commented on src/main/java/com/app/onlinebookstore/service/impl/UserServiceImpl.java:
```diff
@@ -32,6 +36,17 @@ public UserDto register(UserRegisterRequestDto userRegisterRequestDto)
         user.setRoles(Set.of(role));
-        return userMapper.toDto(userRepository.save(user));
+        User savedUser = userRepository.save(user);
+        ShoppingCart shoppingCart = new ShoppingCart();
```
I suggest to extract a logic related to shopping cart to a private method

vovasalyha reviewed (changes requested):
GJ. Let's improve it a bit

Dimagaa committed:
Move ShoppingCart creation logic to ShoppingCartService
Fix issues

andrii-hoienko commented on src/main/java/com/app/onlinebookstore/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,72 @@
+    @PreAuthorize("hasRole('USER')")
+    @PutMapping("/cart-items/{cartItemId}")
+    public CartItemDto updateCartItem(@PathVariable Long cartItemId,
+                               @RequestBody CartItemUpdateRequestDto request) {
```
```suggestion
                               @RequestBody @Valid CartItemUpdateRequestDto request) {
```

andrii-hoienko reviewed (changes requested):
GJ, I left minor comments

Dimagaa committed:
Fix issues
